### PR TITLE
HAL_ChibiOS: prevent long timeouts in DShot

### PR DESCRIPTION
**Pull request by tridge**
this prevents bad calculated timeouts in DShot. The timeout would
sometimes come out as 0xFFFFFFFF, which led to an assert and could
block the thread

This fix is meant to be minimilistic to allow it to be merged easily
into 4.2. A better fix would fix all the uint32_t wrap handling in
DShot

### CODE REVIEW

**Comment by andyp1per** [comment]
Closing in favour of https://github.com/ArduPilot/ardupilot/pull/20430

**Comment by tridge** [comment]
@andyp1per re-opened, as I really want a much simpler change for 4.2, not a dshot rewrite

File: libraries/AP_HAL_ChibiOS/EventSource.cpp [hunk 1]
```diff
@@ -13,7 +14,7 @@ bool EventSource::wait(uint64_t duration, AP_HAL::EventHandle *evt_handle)
     if (duration == 0) {
         ret = chEvtWaitAnyTimeout(evt_mask, TIME_IMMEDIATE);
     } else {
-        ret = chEvtWaitAnyTimeout(evt_mask, chTimeUS2I(duration));
+        ret = chEvtWaitAnyTimeout(evt_mask, chTimeUS2I(MAX(CH_CFG_ST_TIMEDELTA,duration)));
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/EventSource.cpp:17]: Is CH_CFG_ST_TIMEDELTA in ticks or us?

File: libraries/AP_HAL_ChibiOS/EventSource.cpp [hunk 2]
```diff
@@ -13,7 +14,7 @@ bool EventSource::wait(uint64_t duration, AP_HAL::EventHandle *evt_handle)
     if (duration == 0) {
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/EventSource.cpp:14]: feels like you should do < CH_CFG_ST_TIMEDELTA here

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 3]
```diff
@@ -253,14 +253,22 @@ void RCOutput::dshot_collect_dma_locks(uint32_t time_out_us)
...
-                    chTimeUS2I(group.dshot_pulse_send_time_us - pulse_elapsed_us));
+                wait_us = group.dshot_pulse_send_time_us - pulse_elapsed_us;
             }
+
+            // waiting for a very short period of time can cause a
+            // timer wrap with ChibiOS timers. Use CH_CFG_ST_TIMEDELTA
+            // as minimum. Don't allow for a very long delay (over 1200us)
+            // to prevent bugs in handling timer wrap
+            const uint32_t max_delay_us = 1200;
+            wait_us = constrain_uint32(wait_us, CH_CFG_ST_TIMEDELTA, max_delay_us);
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:269]: so I think the correct value here is _dshot_period_us (rather than 1200) - we should try that first

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 3 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp:269]: i don't think so - we're interested in the max time we're willing to wait, which could be longer than dshot period with DMA contention

File: libraries/AP_HAL_ChibiOS/EventSource.cpp (hunk 2 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/EventSource.cpp:14]: duration==0 is part of the API, so it needs to be a special case of no-wait

File: libraries/AP_HAL_ChibiOS/EventSource.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_HAL_ChibiOS/EventSource.cpp:17]: its ticks, but on all but a few peripheral boards on F1 ticks==microseconds

**Comment by tridge** [comment]
merged as part of #20442

**Pull request closed without merging**